src/NHibernate/Hql/Ast/ANTLR/HqlSqlWalker.cs

```diff
@@ -1195,14 +1256,7 @@ public void Visit(IASTNode node)
 			{
 				DotNode dotNode = ( DotNode ) node;
 				FromElement fromElement = dotNode.FromElement;
-				if ( _referencedFromElement != null )
-				{
-					if ( fromElement != _referencedFromElement ) 
```

Comment:
It seems this check was completely removed to also fix the following hibernate [issue](https://hibernate.atlassian.net/browse/HHH-7321). I've added `CrossJoinAndWithClause` to simulate it and this exception is indeed need to be removed. But this test still fails due some other issue (theta style join is simply skipped in generated SQL).